csrc/fusion_segmenter.cpp

```diff
@@ -4206,6 +4207,124 @@ void SegmentCandidateFinder::findSegments() {
+        continue;
+      }
+
+      // Check if the input is an output of an upcast op
+      auto maybe_upcast_op =
+          dynamic_cast<UnaryOp*>(maybe_upcast_out_tv->definition());
+      if (maybe_upcast_op == nullptr ||
+          maybe_upcast_op->getUnaryOpType() != UnaryOpType::Cast) {
+        continue;
+      }
+
+      auto precisions =
+          ir_utils::getPrecisionOfProducerConsumerTensors(maybe_upcast_op);
+      if (!precisions.has_value() || precisions->first >= precisions->second) {
+        continue;
+      }
+
+      // Check if there's multiple uses of the upcast output
+      auto uses_of_upcast_out_tv = maybe_upcast_out_tv->uses();
+      if (uses_of_upcast_out_tv.size() < 2) {
+        continue;
+      }
+
+      // If this is the first use of the upcast output, keep it as is
+      if (expr == uses_of_upcast_out_tv.front()) {
+        continue;
+      }
+
+      auto upcast_out_tv_clone =
+          castOp(maybe_upcast_out_tv->dtype(), maybe_upcast_op->input(0));
+      std::cerr << "Recomputed cast: "
+                << upcast_out_tv_clone->definition()->toString();
+      expr = ir_utils::replaceValInExprInputs(
+          expr, maybe_upcast_out_tv, upcast_out_tv_clone);
+
+      privatized_upcast_ops_[maybe_upcast_op].insert(
+          upcast_out_tv_clone->definition()->as<UnaryOp>());
+    }
+  }
+}
+
+void SegmentCandidateFinder::revertPrivatizedUpcast(SegmentedGroup* group) {
+  for (const auto& [original_upcast, clones] : privatized_upcast_ops_) {
+    std::vector<UnaryOp*> upcast_in_group;
+    Val* upcast_val_to_keep = nullptr;
+    for (auto uop : ir_utils::filterByType<UnaryOp>(group->exprs())) {
+      if (uop != original_upcast && !clones.count(uop)) {
+        continue;
+      }
+
+      auto upcast_tv = uop->out();
+
+      // Prefer the original upcast if found
+      if (upcast_val_to_keep == nullptr ||
+          upcast_tv == original_upcast->out()) {
+        upcast_val_to_keep = upcast_tv;
+      }
+    }
+
+    if (upcast_in_group.size() < 2) {
+      continue;
+    }
+
+    for (const auto i : c10::irange(upcast_in_group.size())) {
+      UnaryOp* uop = upcast_in_group.at(i);
+      Val* upcast_val_to_replace = uop->out();
+      if (upcast_val_to_replace == upcast_val_to_keep) {
+        // Keep this uop as is since its output replaces the other
+        // upcast outputs
+        continue;
+      }
+
+      // Replace upcast_val_to_replace with upcast_val_to_keep
+      for (auto& expr : group->exprs_) {
+        auto input_it = std::find(
+            expr->inputs().begin(),
+            expr->inputs().end(),
+            upcast_val_to_replace);
+        if (input_it == expr->inputs().end()) {
+          continue;
+        }
+
+        auto updated_expr = ir_utils::replaceValInExprInputs(
+            expr, *input_it, upcast_val_to_keep);
+        expr = updated_expr;
+
+        // Update consumer edge vals
+        for (auto consumer_edge : group->consumer_edges) {
```

Comment:
this is inside the loop `for (auto& expr : group->exprs_)` and it doesn't feel right.

Reply:
would we ended up with duplicated edge after this one? If segmentation happens after the privatized cast and we try to merge them together.

e.g.
group0 contains
tv0 -> cast -> tv1 (original)
      -> cast -> tv2  -> used by group1
      -> cast -> tv3  -> used by group1

so there are two edges going from group0 to group1.
But if we merge tv2 and tv3 here, we ended up will only having a single edge.

I'm not sure if the scenario would happen, since we have `MergeUpAndDownCast`, so likely segmenter wouldn't break the upcast from its consumers.

Reply:
also, if we are updating group->consumer_edges, should we worry about the producer_edges of it's consumer groups?

Reply:
> this is inside the loop for (auto& expr : group->exprs_) and it doesn't feel right.
Fixed

Reply:
> But if we merge tv2 and tv3 here, we ended up will only having a single edge.

I think there should be two edges, one for `tv2` and another for `tv3`. So, I think what would happen is both of the two edges would end up having `tv2` as their vals, which is redundant.

I think having redundant edges should be fine, but it could be confusing. I'll update the code to remove those redundant edges.

Reply:
> also, if we are updating group->consumer_edges, should we worry about the producer_edges of it's consumer groups?

No, I don't think so. Edges are shared between producer and consumer segments, so the consumer groups should see the same updated vals.

Reply:
^^ ha I didn't know that. Thx

Reply:
> I'll update the code to remove those redundant edges.

I added a cleanup to remove redundant edges, but I couldn't come up with a test case to trigger this condition. We need segmentation right after two upcast ups and both of the upcast outputs must be used by the same segment. Tried to repro with segment_set, but wasn't successful.

Reply:
Is this because `segment_set` segment on the output of the node?

Reply:
Not sure exactly. I may be missing something. Happy to add a repro if you could create!